### PR TITLE
MSC2996: Add t-shirt

### PR DESCRIPTION
**Pull request by Sorunome**
[Rendered](https://github.com/Sorunome/matrix-doc/blob/soru/t-shirt/proposals/2996-tshift.md)

Please note that this proposal is meant as a joke and is not to be taken seriously

### CODE REVIEW

**Comment by Cadair** [comment]
But what colour should they be?

**Comment by SimonBrandner** [comment]
And what sizes will be available?

**Comment by SimonBrandner** [comment]
> And what sizes will be available?

Hmmm, I think we should also use something better than S, M, L and Xl... We should use some units like beard-seconds! Also, we should probably display the sizes in binary for simplicity

**Comment by PaarthShah** [comment]
From a purely grammatical nitpick, I have to point out that even assuming that the Matrix spec will never be complete and that there will always be more MSCs, that MSC2997 as-rendered will never be able to refer to itself.

That is to say, in the sentence "There's an MSC for *that*!" (emphasis mine), present *on* the T-shirt, it would be improper grammar for "that" to be allowed to refer to MSC2997, and thus explain the T-Shirt's existence.

Potential solutions:
Add a second line: "...and an MSC for this." ("this" being the operative reflexive pronoun to properly reference MSC2997)
OR
Disambiguate the original phrase: "If you can think of it, there's an MSC for it!"

Potential downsides of either include "ruining the joke", but I think that's a perfectly acceptable compromise for technical correctness.

**Comment by MTRNord** [comment]
> From a purely grammatical nitpick, I have to point out that even assuming that the Matrix spec will never be complete and that there will always be more MSCs, that MSC2997 as-rendered will never be able to refer to itself.
> 
> That is to say, in the sentence "There's an MSC for _that_!" (emphasis mine), present _on_ the T-shirt, it would be improper grammar for "that" to be allowed to refer to MSC2997, and thus explain the T-Shirt's existence.
> 
> Potential solutions:
> Add a second line: "...and an MSC for this." ("this" being the operative reflexive pronoun to properly reference MSC2997)
> OR
> Disambiguate the original phrase: "If you can think of it, there's an MSC for it!"
> 
> Potential downsides of either include "ruining the joke", but I think that's a perfectly acceptable compromise for technical correctness.

What about we strikte "that" in the Text and put a handwritten "this" over it to emphanzise how the spec progress works?

File: proposals/2997-tshift.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# MSC2997: "There's an MSC for that!"
```
Comment (Sorunome) [inline proposals/2997-tshift.md:1]: Should we consider t-shirt's-as-rooms?

File: proposals/2997-tshift.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+Once the spec is complete and there are no more MSCs the t-shirt will be obsolete.
+
+## Alternatives
+
+Instead of a shirt, a sweater or thelike could be used.
+
+## Security considerations
+
+Unfortunately not all people are able to wear t-shirts, so a solution for those would have to be
```
Comment (MTRNord) [inline proposals/2997-tshift.md:21]: what about wristbands with text for them?

**Comment by deepbluev7** [comment]
Please comment on some line of the MSC, since otherwise the discussion is hard to follow, since you don't get threads!

**Comment by ShadowJonathan** [comment]
In all seriousness, i'd like to have one on https://shop.matrix.org, preferably in the following format;

`[matrix]` logo on top, and underneath, in hand-writing font; `"There's an MSC for that!"`

**Comment by MTRNord** [comment]
> In all seriousness, i'd like to have one on https://shop.matrix.org, preferably in the following format;
> 
> `[matrix]` logo on top, and underneath, in hand-writing font; `"There's an MSC for that!"`

Use threads please

File: proposals/2997-tshift.md (hunk 2 above)
Comment (anoadragon453) [reply proposals/2997-tshift.md:21]: This is a fair point, though I'm not sure if it would mitigate the security that a t-shirt provides. For one, the surface area of a t-shirt far outweighs that of a wristband. Wristbands can also be made from a variety of materials, some even opting for mere paper.

Not to mention the lack of area for displaying the intended messaging of the MSC.

I think if wristbands is taken as an approach, care should be taken to ensure the security and practicality of the device.

Comment (MTRNord) [reply proposals/2997-tshift.md:21]: What about a hat? Or similar headware? It is more visible than a wristband and everyone that is affected by MSC tends to have a head. Or do we also want to integrate aliens via a bridge end expect headless users?

Comment (deepbluev7) [reply proposals/2997-tshift.md:21]: What about capes? Almost everyone has a neck and you can make them pretty big to put arbitrary messages on them. Would need to come with a disclaimer not to wear them next to aircrafts though.

**Comment by uhoreg** [comment]
> Use threads please

Yes, threads are important to use for T-shirts

File: proposals/2997-tshift.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
+# MSC2997: "There's an MSC for that!"
+
+Throughout fosdem 2021 multiple different features or ideas around matrix were discussed. As it turns
+out, most of them already had an MSC for that! One feature that was keep being brought up, was to get
+"There's an MSC for that!" onto a t-shirt. Unfortunately, there wasn't an MSC for that yet.
+
+## Proposal
+
+Get a t-shirt with "There's an MSC for that!" written on it.
```
Comment (erkinalp) [inline proposals/2997-tshift.md:9]:
```suggestion
Get a t-shirt with "There's an MSC for that!" written on it. This T-shirt shall include a code block typeset in a monospace block that
\```
{
   type: "org.matrix.t-shirt",
   content: "There's an MSC for that"
}
```
```

**Review by erkinalp — CHANGES_REQUESTED** [review]

File: proposals/2997-tshift.md (hunk 3 above)
Comment (Sorunome) [reply proposals/2997-tshift.md:9]: Good point

File: proposals/2997-tshift.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+## Proposal
+
+Get a t-shirt with "There's an MSC for that!" written on it. This T-shirt shall include a code block
+typeset in a monospace block that should read
+```json
+{
+  "type": "m.t-shirt",
+  "content": {
+    "body": "There's an MSC for that!"
```
Comment (0xADD1E) [inline proposals/2997-tshift.md:15]: Should `body` in this case actually be the user? T-shirts are generally most useful on the user's body

Comment (Sorunome) [reply proposals/2997-tshift.md:15]: As @uhoreg pointed out [here](https://github.com/matrix-org/matrix-doc/pull/2241#discussion_r542538401) it might be a plan that, moving forward, any event should have a fallback `body` to display.

Comment (0xADD1E) [reply proposals/2997-tshift.md:15]: My suggestion is simply that the t-shirt will be more useful if the body in question were to be the user's body instead of some arbitrary text. 

Properly serialising the end user may however prove to be a significant implementation challenge

File: proposals/2997-tshift.md (hunk 1 above)
Comment (Sorunome) [reply proposals/2997-tshift.md:1]: Added this

File: proposals/2997-tshift.md (hunk 4 above)
Comment (Sorunome) [reply proposals/2997-tshift.md:15]: Oooh, do you mean like this?

```json
{
  "type": "m.t-shirt",
  "content": {
    "body": "@sorunome:sorunome.de"
  }
}
```

Comment (kate-shine) [reply proposals/2997-tshift.md:15]: That's a medical stuff, so it's probably depending on us to spec HL7 FHIR resource state events. ^_^

Comment (Sorunome) [reply proposals/2997-tshift.md:15]: Maybe we should make it a state event, then? Since state_keys starting with `@` are preserved so that they only can be set by the user themself and not by others.

Comment (kate-shine) [reply proposals/2997-tshift.md:15]: Yes! And the content is quite obvious. 
https://www.hl7.org/fhir/genomics.html

File: proposals/2997-tshift.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+```
+## Potential issues
+
+Once the spec is complete and there are no more MSCs the t-shirt will be obsolete.
+
+## Alternatives
+
+Instead of a shirt, a sweater or thelike could be used.
+
+Instead of just printing normal event contents on the shirt we might want to consider using
```
Comment (0xADD1E) [inline proposals/2997-tshift.md:27]: While the increased flexibility for metadata would be useful, specific care should be given to the sizing implementation. Smaller users may have issues with the fit of a room based t-shirt

Comment (kate-shine) [reply proposals/2997-tshift.md:27]: It could allow us to have not only "One size fits all" t-shirts, but also "One size fits many" - at once!

File: proposals/2997-tshift.md (hunk 4 above)
Comment (kate-shine) [reply proposals/2997-tshift.md:15]: On the other hand, I'm quite worried that the serialization of person would require a space of t-shirt-as-rooms

**Review by erkinalp — APPROVED** [review]

**Review by ShadowJonathan — APPROVED** [review]

File: proposals/2997-tshift.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+out, most of them already had an MSC for that! One feature that was keep being brought up, was to get
+"There's an MSC for that!" onto a t-shirt. Unfortunately, there wasn't an MSC for that yet.
+
+## Proposal
+
+Get a t-shirt with "There's an MSC for that!" written on it.
+
+## Potential issues
+
+Once the spec is complete and there are no more MSCs the t-shirt will be obsolete.
```
[the remark this answers is not in the document]
Comment (erkinalp) [reply proposals/2997-tshift.md:21]: > 
```suggestion
Once the spec is complete and there are no more MSCs the t-shirt will be obsolete.
That is never going to happen, though.
```

File: proposals/2997-tshift.md [hunk 7]
```diff
@@ -0,0 +1,39 @@
+# MSC2997: "There's an MSC for that!"
+
+Throughout fosdem 2021 multiple different features or ideas around matrix were discussed. As it turns
+out, most of them already had an MSC for that! One feature that was keep being brought up, was to get
+"There's an MSC for that!" onto a t-shirt. Unfortunately, there wasn't an MSC for that yet.
+
+## Proposal
+
+Get a t-shirt with "There's an MSC for that!" written on it. This T-shirt shall include a code block
+typeset in a monospace block that should read
```
Comment (erkinalp) [inline proposals/2997-tshift.md:10]:
```suggestion
Get a t-shirt. This T-shirt shall include a code block typeset in a monospace block that should read
```

**Review by erkinalp — CHANGES_REQUESTED** [review]

File: proposals/2997-tshift.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  }
+}
+```
+## Potential issues
+
+Once the spec is complete and there are no more MSCs the t-shirt will be obsolete.
+
+## Alternatives
+
+Instead of a shirt, a sweater or thelike could be used.
```
Comment (SimonBrandner) [inline proposals/2997-tshift.md:25]: It seems shirts have not been considered which I find quite sad since I am a regular user of them

File: proposals/2997-tshift.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
+# MSC2997: "There's an MSC for that!"
```
Comment (MTRNord) [inline proposals/2997-tshift.md:1]: Mostly for visual reference based on the current MSC state. :) Not finished in the slightest, and more a 15m photoshop adventure than anything else. But maybe it helps future readers to visualize some ideas on this MSC.
Front

**Comment by yncyrydybyl** [comment]
There is another draft for an implementation from @nadineminagawa. It will cover the front end and back end implementation of this MSC. 
The fake handwriting should be replaced by some real handwriting. And we need to find out if we want to have the side effects.
[MSC-Shirt.pdf](https://github.com/matrix-org/matrix-spec-proposals/files/12072508/MSC-Shirt.pdf)

**Comment by HarHarLinks** [comment]
> [MSC-Shirt.pdf](https://github.com/matrix-org/matrix-spec-proposals/files/12072508/MSC-Shirt.pdf)

- please note that as long as this MSC isn't merged, the unstable prefix `de.sorunome.t-shirt` should be used for the backend implementation as demonstrated by MTRNord ;)
- I'm not sure why the n on the frontend would be crossed out, but then I'm not a native English speaker
image

**Comment by yncyrydybyl** [comment]
> I'm not sure why the n on the frontend would be crossed out, but then I'm not a native English speaker

"_an_" is used if the following words sounds like a vowel but "_a_" for a consonant sound.

We can stay with the "_a_". There is the example of an MP3-Player at https://dictionary.cambridge.org/grammar/british-grammar/a-an-and-the

image

**Comment by HarHarLinks** [comment]
> "an" is used if the following words sounds like a vowel but "a" for a consonant sound.

exactly, "MSC" is pronounced ~ "emm ess see", which means it should be "an MSC" same as "an MP3"

**Comment by axelsimon** [comment]
> There is another draft for an implementation from @nadineminagawa. It will cover the front end and back end implementation of this MSC.

Quick nitpick: the apostrophe used on the t-shirt draft is upside down, it should be `’` (ie, heavy part up, cf. [wikipedia article](https://en.wikipedia.org/wiki/Apostrophe)).

File: proposals/2997-tshift.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+Get a t-shirt with "There's an MSC for that!" written on it. This T-shirt shall include a code block
+typeset in a monospace block that should read
+```json
+{
+  "type": "m.t-shirt",
+  "content": {
+    "body": "There's an MSC for that!"
+  }
+}
```
Comment (HarHarLinks) [inline proposals/2997-tshift.md:17]: A POC implementation was done for [CCCamp23](https://events.ccc.de/camp/2023/)!

The following constitutes the backend implementation
```suggestion
{
  "type": "m.t-shirt",
  "content": {
    "body": "There's an MSC for that!",
    "background": "$bgcolor",
    "text-color": "$fgcolor",
    "size": "$size",
    "fitted": $fitted,
    "frontend.logo": $logo,
    "backend.highlight": $highlight
  },
  "origin_server_ts": $timestamp
}
```
where
- bgcolor is the shirt's color, in this case `#000000`
- fgcolor is the frontend print's color, in this case `#FFFFFF`
- size is the T-shirt size in T-Shirt sizes, we used `S`, `M`, `L`, `XL`, `XXL`, `XXXL`
- fitted indicates whether it's a fitted T-shirt (boolean)
- logo is whether the frontend depicts the `[matrix]` logo (boolean), in this case `true`
- highlight is whether the backend implementation uses syntax highlighting (boolean), in this case `true`
- timestamp is the creation or event start, in this case `1692049320000`

Full event as implemented by us:
```
{
  "type": "de.sorunome.t-shirt",
  "org.matrix.msc2997.content": {
    "body": "There's an MSC for that!",
    "background": "#000000",
    "text-color": "#FFFFFF",
    "size": "M",
    "fitted": true,
    "frontend.logo": true,
    "backend.highlight": true
  },
  "origin_server_ts": 1692049320000
}
```

**Review by HarHarLinks — CHANGES_REQUESTED** [review]

File: proposals/2997-tshift.md (hunk 10 above)
Comment (HarHarLinks) [reply proposals/2997-tshift.md:17]: (gratz @Sorunome, you're eternalised in a couple T-shirts now.... :tada:)

Comment (turt2live) [reply proposals/2997-tshift.md:17]: (implementation accepted - MSC description updated)

Comment (HarHarLinks) [reply proposals/2997-tshift.md:17]: We iterated for [Matrix Community Summit 2023](https://summit2023.matrixmeetup.de/conference/)!
It is based on the [extensible events](https://github.com/matrix-org/matrix-spec-proposals/pull/1767) for [media](https://github.com/matrix-org/matrix-spec-proposals/pull/3552) and by extension, [files](https://github.com/matrix-org/matrix-doc/pull/3551).
`content` contains 3 blocks:
- Like the media extensible events, there is the `m.text` fallback block in case your t-shirt manufacturer is unable to reproduce the graphic given by the file.
- The 'm.file` block holds some common properties
- The `m.t-shirt` block (unstable prefix `org.matrix.msc2997.t-shirt`) specifies some t-shirt specific details

Some details of the event metadata are not shown in the backend for comprehensiveness.

```suggestion
{
  "type": "de.sorunome.t-shirt",
  "sender": "$sender_mxid",
  "content": {
    "m.text": [
      { "body": "There's an MSC for that!" }
    ],
    "m.file": {
      "url": "mxc://$some_url",
      "name": "$url_label",
      "mimetype": "$mime",
      "size": "$size"
    },
    "org.matrix.msc2997.t-shirt": {
      "background": "$bgcolor",
      "text-color": "$fgcolor",
      "fitted": $fitted,
      "frontend.logo": $logo
      "backend.highlight": $highlight
    }
  },
  "origin_server_ts": $event_start
}
```
where:
- `sender_mxid` per https://spec.matrix.org/latest/client-server-api/#room-event-format
- `some_url` a url in the format of https://spec.matrix.org/latest/client-server-api/#matrix-content-mxc-uris that coincidentally also works when using it with the HTTP(s) protocol instead of MXC - we used `mxc://summit2023.matrixmeetup.de/conference`
- `mime` the material of the shirt, in this case `t-shirt/cotton`
- `size` is the T-shirt size in T-Shirt sizes, we used `S`, `M`, `L`, `XL`, `XXL`, up to `5XL`
- `bgcolor` is the shirt's color, in this case `#000000`
- `fgcolor` is the frontend print's color, in this case `#FFFFFF`
- optionally, `fitted` indicates whether it's a fitted T-shirt (boolean) - our implementation lists it explicitly even if false
- `logo` is whether the frontend depicts the graphic indicated by the `m.file.url` (boolean), in this case `true`
- `highlight` is whether the backend implementation uses syntax highlighting (boolean), in this case `true`
- `timestamp` is the creation or event start, in this case `1695283200000`

Full event as implemented by us:
```jsonc
{
  "type": "de.sorunome.t-shirt",
  "sender": "@me:homeserver.tld",
  "content": {
    "m.text": [
      { "body": "There's an MSC for that!" }
    ],
    "m.file": {
      "url": "mxc://summit2023.matrixmeetup.de/conference",
      "name": "Matrix Community Summit 2023",
      "mimetype": "t-shirt/cotton",
      "size": "L"
    },
    "org.matrix.msc2997.t-shirt": {
      "background": "#000000",
      "text-color": "#FFFFFF",
      "fitted": true,
      "frontend.logo": true,
      "backend.highlight": true
    }
  },
  "origin_server_ts": 1695283200000
}
```
